src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_AL_SIZE = 10;
+    private static final double CAPACITY_GROWTH_VALUE = 1.5;
+    private static final int ZERO_NUMBER = 0;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_AL_SIZE];
+        size = ZERO_NUMBER;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add value by this index");
+        }
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] array = (T[]) new Object[list.size()];
+        for (int i = 0; i < array.length; i++) {
+            array[i] = list.get(i);
+            add(array[i]);
+        }
     }
```

Comment:
The 'addAll' method can be optimized by ensuring the capacity once before the loop instead of calling 'ensureCapacity' inside the loop on each 'add' operation.